MHCFlurry and MHCNuggets Interfaces

Zethson opened the pull request:
Hi,

this PR solves https://github.com/FRED-2/Fred2/issues/224 .

A few notes about the implementation:
Both tools are somewhat verbose, hence it would make sense to silence them (they don't offer that via their python API). This is far easier in python 3 (context switching), hence I suggest that this is done after python3 is finally supported. It's not that bad, so for now it's surely fine.
Secondly, even though they may not be real PSSM methods, they're most easily implemented via the PSSM interface. One could think about adding another interface for methods that are python packages to prevent the PSSM interface from getting cluttered by too many methods.

1. MHCFlurry
MHCFlurry requires the models to be downloaded and accessible from the filesystem. I implemented this in the following way: If the models are not accessible, they will be downloaded and are by default saved to ~/.local/mhcflurry/[...] .
From the MHCFlurry doc: "Files downloaded with mhcflurry-downloads are stored in a platform-specific directory. " They don't seem to offer any alternatives.

2. MHCNuggets
MHCNuggets does for some reason only allow files as input and output for their python API. Hence, I save the input peptides as files in /tmp , predict them, output them as a file in /tmp and finally read them back in. Not beautiful, but it works.

3. Code can be tested via
```
peptides = [Peptide("SYFPEITHI"), Peptide("FIASNGVKL"), Peptide("LLGATCMFV")]
    print EpitopePredictorFactory("mhcflurry").predict(peptides)
    print EpitopePredictorFactory("mhcnuggets-class-1").predict(peptides)
    print EpitopePredictorFactory("mhcnuggets-class-2").predict(peptides)
```

Zethson commented:
Please review @b-schubert @christopher-mohr @Leon-Bichmann .

b-schubert commented:
@Zethson Thanks so much for doing this! I will have a closer look as soon as I have a bit of spare time.

Two things I noticed while briefly going through the PR:

1) APSSMEpitopePrediction might not be the best interface for the two ANN-based predictors since this interface is used in other modules to filter for linear predictors that are embedded into algorithms that require linearity. I would rather specify a new interface as you suggested.

2) The MHC-II allele strings are not in the right format. See the NetMHCpan-II implementation for the correct format.

Zethson commented:
@b-schubert Hope that I've addressed your points.
ANN interface added, inspired by the PSSM interface.
MHC-II alleles should now be in the right format.

b-schubert commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -0,0 +1,455 @@
+        Implements predict functionality
+    """
+
+    def predict(self, peptides, alleles=None, **kwargs):
```
Would it make sense to raise a NotImplementedError and have the downstream interfaces implement this function since potentially every toll will have its custom ```predict``` function?

Zethson replied:
Added a metaclass which also checks for abstract method parameter signature violation.

b-schubert commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -0,0 +1,455 @@
+         "HLA-B51:02", "HLA-B52:01", "HLA-B53:01", "HLA-B54:01", "HLA-B55:01", "HLA-B55:02", "HLA-B56:01",
+         "HLA-B57:01", "HLA-B57:02", "HLA-B57:03", "HLA-B58:01", "HLA-B58:02", "HLA-B60:01", "HLA-B61:01",
+         "HLA-B62:01", "HLA-B73:01", "HLA-B81:01", "HLA-B83:01"])
+    __supported_length = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]
```
should be a frozenset

b-schubert commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -0,0 +1,455 @@
+                           "HLA-DPB1*03:01", "HLA-DPB1*04:01", "HLA-DPB1*04:02", "HLA-DPB1*05:01", "HLA-DPB1*09:01",
+                           "HLA-DPB1*11:01", "HLA-DPB1*14:01", "HLA-DPB1*20:01", "HLA-DQA1*01:01",
+                           "HLA-DQA1*01:01-DQB10501", "HLA-DQA1*01:01-DQB1*05:03", "HLA-DQA1*01:02",
+                           "HLA-DQA1*01:02-DQB1*05:01", "HLA-DQA1*01:02-DQB10502", "HLA-DQA1*01:02-DQB1*06:02",
```
Some of the DQB1 alleles do not follow the naming convention

b-schubert commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -0,0 +1,455 @@
+                           "HLA-DQA1*03:02-DQB10301", "HLA-DQA1*03:02-DQB1*03:03", "HLA-DQA1*03:02-DQB10401",
+                           "HLA-DQA1*03:03-DQB10402", "HLA-DQA1*04:01-DQB1*04:02", "HLA-DQA1*05:01",
+                           "HLA-DQA1*05:01-DQB1*02:01", "HLA-DQA1*05:01-DQB1*03:01", "HLA-DQA1*05:01-DQB1*03:02",
+                           "HLA-DQA1*05:01-DQB10303", "HLA-DQA1*05:01-DQB10402", "HLA-DQA1*05:05-DQB1*03:01",
```
naming convention

b-schubert commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -0,0 +1,455 @@
+                           "HLA-DRB1*13:05", "HLA-DRB1*14:01", "HLA-DRB1*14:02", "HLA-DRB1*15:01", "HLA-DRB1*15:02",
+                           "HLA-DRB1*15:03", "HLA-DRB1*16:01", "HLA-DRB1*16:02", "HLA-DRB3*01:01", "HLA-DRB3*02:02",
+                           "HLA-DRB3*03:01", "HLA-DRB4*01:01", "HLA-DRB4*01:03", "HLA-DRB5*01:01", "HLA-DRB5*01:02"])
+    __supported_length = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]
```
frozenset

b-schubert commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -0,0 +1,455 @@
+import pandas
+
+from mhcflurry import Class1AffinityPredictor
+from mhcnuggets.src.predict import predict
```
Would be good either use just ```import mhcnugget```  or to rename ```predict```  to ```mhcnugget_predict``` as not to confuse who's  ```predict``` function is called once more than just the two models have been integrated.

b-schubert commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -0,0 +1,455 @@
+        for a in alleles:
+            result[a] = {}
+            tmp_output_file = tempfile.NamedTemporaryFile().name
+            predict(class_='I',
```
See comment in import section

Zethson replied:
@b-schubert I hope that I fixed all class 2 allele formatting stuff now. 

Rest is addressed as well.
I've added a metaclass for abstract method contracts, which even checks for matching parameter signatures (not simple in python).
Hence, the predict method is now completely abstract. All new ANN based predictors should use the metaclass 'SignatureCheckerMeta' now.
I hope that this is what you also had in mind.

Zethson commented:
I hope that I've caught all class 2 allele format issues now.

I've also addressed the other stuff. I've added a metaclass which allows for abstract method + method signature checking.
=> If a method does not adhere to the parameter contract, an exception will be thrown. This is not trivial in python, but the metaclass allows us to do so.
=> All ANN based predictors have to use the metaclass now and predict is completely abstract in ANN.py

I hope that this is what you also had in mind.

Zethson commented:
Instead of ugly prints I could also make use of modern logging, but since this is not yet used throughout the project, it may violate the style of the project.

I don't want to solve the logging issue in this PR, since this is another issue, but I can do so if you want me to.

b-schubert commented:
@Zethson Thanks again for the PR. Looks great now!

Regarding logging and printing. Yes that is absolutely something we wanted to change. There is even an Issue dedicated to that. I you want you can absolutely go ahead and implement proper logging!
